components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt

```diff
@@ -433,37 +458,23 @@ class PromptFeature(
         }
     }
 
-    /**
-     * Observes [Session.Observer.onPromptRequested] of the selected session
-     * and notifies the feature whenever a prompt needs to be shown.
-     */
-    internal class PromptRequestObserver(
-        sessionManager: SessionManager,
-        private val feature: PromptFeature
-    ) : SelectionAwareSessionObserver(sessionManager) {
-
-        override fun onPromptRequested(session: Session, promptRequest: PromptRequest): Boolean {
-            feature.onPromptRequested(session, promptRequest)
-            return false
-        }
-
-        override fun onLoadingStateChanged(session: Session, loading: Boolean) {
-            if (!loading) {
-                feature.promptAbuserDetector.resetJSAlertAbuseState()
-            }
-        }
-    }
-
     companion object {
         const val FILE_PICKER_ACTIVITY_REQUEST_CODE = 1234
     }
 }
 
-internal fun SessionManager.consumePromptFrom(
+internal fun BrowserStore.consumePromptFrom(
     sessionId: String?,
-    consumer: (PromptRequest) -> Boolean
+    consume: (PromptRequest) -> Unit
 ) {
-    runWithSessionIdOrSelected(sessionId) {
-        it.promptRequest.consume(consumer)
+    if (sessionId == null) {
+        state.selectedTab
+    } else {
+        state.findTabOrCustomTab(sessionId)
+    }?.let { tab ->
+        tab.content.promptRequest?.let {
+            consume(it)
+            dispatch(ContentAction.ConsumePromptRequestAction(tab.id))
+        }
```

Comment:
Don't we still need to consume from `Session(Manager)` since otherwise both are not in sync anymore? That's where we used UseCases in other situations.

At least unless we do not remove it from `Session`, which we could do too since I'm pretty sure no one else except this feature reads prompt requests. Otherwise this would be a good follow-up.

In this current version I fail to understand why we try to get the tab first and then dispatch with its id when we have that id already passed to this method?

Reply:
@pocmo 

> In this current version I fail to understand why we try to get the tab first and then dispatch with its id when we have that id already passed to this method?

That's because we need to get to the prompt request to consume it. So, we need to lookup the "tab" to get to its `content.promptRequest`.


>At least unless we do not remove it from Session, which we could do too since I'm pretty sure no one else except this feature reads prompt requests. Otherwise this would be a good follow-up.

Ah, you're right! I've added a commit here to remove the property from browser session. This way we don't have to keep it in sync and nobody else uses it.